src/pip/_internal/resolution/resolvelib/requirements.py

```diff
@@ -97,15 +103,40 @@ def find_matches(self):
                 ican.link,
                 self._preparer,
                 self._ireq,
-                self._make_install_req
+                self._make_install_req,
+                self._make_python_req,
             )
             for ican in found.iter_applicable()
         ]
 
     def is_satisfied_by(self, candidate):
         # type: (Candidate) -> bool
-
         assert candidate.name == self.name, \
             "Internal issue: Candidate is not for this requirement " \
             " {} vs {}".format(candidate.name, self.name)
         return candidate.version in self._ireq.req.specifier
+
+
+class RequiredPythonRequirement(Requirement):
+    def __init__(self, py_version, specifier):
+        # type: (_BaseVersion, SpecifierSet) -> None
+        self.specifier = specifier
+        self.py_version = py_version
+
+    @property
+    def name(self):
+        # type: () -> str
+        return "<Python>"  # To avoid conflicting with the real PyPI package.
+
+    def find_matches(self):
+        # type: () -> Sequence[Candidate]
+        candidate = RequiredPythonCandidate(self.py_version)
+        if self.is_satisfied_by(candidate):
+            return [candidate]
+        return []
+
+    def is_satisfied_by(self, candidate):
+        # type: (Candidate) -> bool
+        assert isinstance(candidate, self.__class__), \
```

Comment:
`self.__class__` is a `RequiredPythonRequirement`, not a `RequiredPythonCandidate`. Might need the explicit candidate class here (more coupling between requirement and candidate modules 🙁)

Reply:
This could also be `assert candidate.name == "<Python>"` (or is this too magical?)